### ISSUE TITLE
Scene2d label wrapping incorrectly

### ISSUE THREAD

**Issue by lucas-kakele** [libgdx/libgdx#6098]
When using label wrapping, for some texts, the wrapping isn't performed correctly.

![label-wrap-bug](https://user-images.githubusercontent.com/56004737/87057120-d8a6d700-c1dc-11ea-8a06-1250b8cbdc3f.png)

As you can see in the screenshot above, "water." text fits the line. In fact, if you add a blank space to the end (or an extra dot), it draws in a single line.

Code to reproduce:

```
public class LabelWrappingBug extends ApplicationAdapter {

	private BitmapFont bitmapFont;
	private Stage stage;
	
	@Override
	public void create () {
		stage = new Stage(new ScreenViewport());
		bitmapFont = new BitmapFont(Gdx.files.internal("ringbearer.fnt"));

		LabelStyle style = new LabelStyle(bitmapFont, Color.WHITE);
		Label label = new Label("It is filled with water.", style);
		label.setWrap(true);
		label.setAlignment(Align.center);

		Table table = new Table();
		table.add(label).prefWidth(600);

		table.setFillParent(true);

		table.debug();

		stage.addActor(table);
	}

	@Override
	public void render () {
		Gdx.gl.glClearColor(0, 0, 0, 1);
		Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);

		stage.act();
		stage.draw();
	}
	
	@Override
	public void dispose () {
		stage.dispose();
		bitmapFont.dispose();
	}
}
```

Font files:

[ringbearer.zip](https://github.com/libgdx/libgdx/files/4897710/ringbearer.zip)

The LibGDX version I'm using is 1.9.10. I'm testing on Desktop, using Ubuntu 18.04 LTS. Java version is 1.8.0_201.

I believe the issue is explained here: https://stackoverflow.com/questions/43618643/libgdx-label-text-wrap-bug

It has something to do with second call to `layout.setText` in `Label` class, probably a rounding issue.

I suspect this is also causing the labels padding to be calculated wrong sometimes. For example (text is cut of on the right):

![bug-1](https://user-images.githubusercontent.com/56004737/87058190-2a039600-c1de-11ea-9656-4aa3f40b2733.png)
![bug-2](https://user-images.githubusercontent.com/56004737/87058203-2d971d00-c1de-11ea-88bf-15d73cb2c6fd.png)

Let me know if you need more information and if I can help somehow! Kind regards, Lucas.

**Comment by NathanSweet** [comment]
Label does layout twice: first to determine the wrapped width, then again using that width. Maybe this could be avoided, I don't remember why it the first one isn't sufficient. It sure does a good job finding bugs though.

For the first layout, this check:
https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/graphics/g2d/GlyphLayout.java#L180
The glyph `r` (in "water") has a `glyphWidth` of 26 and the `xAdvance` is 17. `x` is 347, so + 17 = 364. The period after `r` is narrow at 7 wide, so 364 + 7 = 371, which is the GlyphLayout width.

The second layout should fit within 371, but 347 + the 26 width of `r` = 373. The `r` is so large it extends past the period, something like this, where the red line is the `xAdvance`:
![](http://n4te.com/x/607-JKzK.png)

This is a discrepancy between the GlyphLayout width and the width which will fit the text without wrapping. Wrapping wants to fit the whole glyph. The bounds are tightly fitting using only the last glyph. I suppose the bounds should use the max of any glyph, which would fix this.

This isn't a rounding problem. If that is an issue maybe we can ceil somewhere, but it hasn't been reported as reproducible.

**Comment by lucas-kakele** [comment]
Thanks Nathan for looking into this. I appreciate your time!

**Comment by NathanSweet** [comment]
Sorry, the fix is not good. The wrapping code modifies the run, so the max width that was stored before that is too large.

**Comment by mttkay** [comment]
Just in case anyone else runs into this: I experienced broken text rendering after bumping libgdx from 1.9.10 to 1.9.11 (this issue I presume?) I was rendering a `TypingLabel` inside a `ScrollPane`, and letters were cut off at the end instead of the word being wrapped into a new line. It also looked like table padding was not taken into account, since text would spill over the padding.

I found that updating to 1.9.12 fixes this problem again, presumably due to this fix:

> - Fixed GlyphLayout not laying out correctly with color markup.

But I'm not sure. I just wanted to mention it because I lost over an hour staring at my own code trying to figure out what I broke when it came to me that I might have pulled in the problem via the version bump that I clearly did not test well enough :-)

Anyway, thanks for putting a fix together so quickly! Appreciate it.
